rfc-drafts/yield_safe_lint.md

```diff
@@ -0,0 +1,89 @@
+# RFC: yield safe lint
+
+# Summary
+
+Introduce a `#[yield_unsafe]` lint in the compiler that will warn the user when they unsafely hold an unsafe yield struct across a yield boundary. 
+
+# Motivation
+
+Enable users to fearlessly write concurrent async code without the need to understand the internals of runtimes and how their code will be affected. The goal is to provide a best effort warning that will let the user know of a possible side effect that is not visible by reading the code right away. Some examples of these side effects are holding a `MutexGuard` across a yield bound in a single threaded runtime. In this case the resulting generated future will resolve to `!Send` but could still hold the lock when the future yield back to the executor. This opens up for the possibility of causing a deadlock since the future holding onto the lock did not relinquish it back before it yielded control.
+
+# Guide-level explanation
+
+Provide a lint that can be attached to structs to let the compiler know that this struct is unsafe to be held across a yield boundary.
+
+```rust
+    #[yield_unsafe]
+    struct MyUnsafeYieldStruct {}
+```
+
+This struct if held across a yield boundary would cause a warning:
+
+```rust
+    async fn foo() {
+      let my_struct = MyUnsafeYieldStruct {};
+      my_async_op.await;
+      println!("{:?}", my_struct);
+    }
+```
+
+The compiler might output something along the lines of:
+
+
+    warning: Holding `MyUnsafeYieldStruct` across the await bound on line 3 might cause side effects.
+
+Examples use cases for this lint:
+
+    - `MutexGuard` holding this across a yield boundary in a single threaded executor could cause deadlocks. In a multi-threaded runtime the resulting future would become `!Send` which will stop the user from spawning this future and causing issues. But in a single threaded runtime which accepts `!Send` futures deadlocks could happen.
+        - The same applies to other such synchronization primitives such as locks from ParkingLot.
+    - `tracing::Span` has the ability to enter the span via the `tracing::span::Entered` guard. While entering a span is totally normal, during an async fn the span only needs to be entered during 
```

Comment:
`during an async fn the span only needs to be entered during`
missing the rest of the sentence